MSAL/src/public/MSALAuthenticationSchemePop.m

```diff
@@ -0,0 +1,134 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#import "MSALAuthenticationSchemePop.h"
+#import "MSIDAuthenticationSchemePop.h"
+#import "MSALHttpMethod.h"
+#import "MSIDDevicePopManager.h"
+#import "MSALAuthScheme.h"
+#import "MSIDAccessToken.h"
+#import "MSIDDefaultTokenCacheAccessor.h"
+
+static NSString *keyDelimiter = @" ";
+
+@interface MSALAuthenticationSchemePop()
+
+@property (nonatomic) NSString *nonce;
+
+@end
+
+@implementation MSALAuthenticationSchemePop
+
+- (instancetype)initWithHttpMethod:(MSALHttpMethod)httpMethod requestUrl:(NSURL *)requestUrl
+{
+    self = [super init];
+    if (self)
+    {
+        _scheme = MSALAuthSchemePop;
+        _httpMethod = httpMethod;
+        _requestUrl = requestUrl;
+        _nonce = [[NSUUID new] UUIDString];
+    }
+
+    return self;
+}
+
+- (id<MSIDAuthenticationSchemeProtocol>)createMSIDAuthenticationSchemeWithParams:(nullable NSDictionary *)params
+{
+    return [[MSIDAuthenticationSchemePop alloc] initWithSchemeParameters:params];
+}
+
+- (NSDictionary *)getSchemeParameters:(MSIDDevicePopManager *)popManager
+{
+    NSMutableDictionary *schemeParams = [NSMutableDictionary new];
+    NSString *requestConf = popManager.requestConfirmation;
+    if (requestConf)
+    {
+        [schemeParams setObject:MSALParameterStringForAuthScheme(self.scheme) forKey:MSID_OAUTH2_TOKEN_TYPE];
+        [schemeParams setObject:requestConf forKey:MSID_OAUTH2_REQUEST_CONFIRMATION];
+    }
+    else
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to append public key jwk to request headers.");
+    }
+    
+    return schemeParams;
+}
+
+- (nullable NSString *)getSecret:(MSIDAccessToken *)accessToken popManager:(nullable MSIDDevicePopManager *)popManager error:(NSError **)error
+{
+    NSString *kid = popManager.kid;
+    if (!kid || !accessToken.kid || ![accessToken.kid isEqualToString:kid])
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to match access token key thumbprint.");
+        BOOL result = [popManager.tokenCache removeToken:accessToken context:nil error:nil];
+        
+        if (!result)
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to remove access token after thumprint mismatch.");
+        }
+        
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Failed to match access token key thumbprint.", nil, nil, nil, nil, nil, YES);
```

Comment:
hmmm, shouldn't we do this matching in MSIDSilentTokenRequest instead such as we would acquire new access token using refresh token? Otherwise, we would block token acquisition here.

Reply:
Good point. Addressed. Added code for public key thumbprint matching similar to enrollment id matching of access tokens